orbit/pkg/installer/installer.go

```diff
@@ -0,0 +1,251 @@
+	installer, err := r.OrbitClient.GetInstallerDetails(installId)
+	if err != nil {
+		return nil, fmt.Errorf("fetching software installer details: %w", err)
+	}
+
+	payload := &fleet.HostSoftwareInstallResultPayload{}
+	payload.InstallUUID = installId
+
+	if installer.PreInstallCondition != "" {
+		shouldInstall, output, err := r.preConditionCheck(ctx, installer.PreInstallCondition)
+		payload.PreInstallConditionOutput = &output
+		if err != nil {
+			return payload, err
+		}
+
+		if !shouldInstall {
+			return payload, nil
+		}
+	}
+
+	if !r.scriptsEnabled() {
+		// fleetctl knows that -2 means script was disabled on host
+		payload.InstallScriptExitCode = ptr.Int(-2)
+		payload.InstallScriptOutput = ptr.String("Scripts are disabled")
+		return payload, nil
+	}
+
+	if r.tempDirFn == nil {
+		r.tempDirFn = os.MkdirTemp
+	}
+	tmpDir, err := r.tempDirFn("", "")
+	if err != nil {
+		return payload, fmt.Errorf("creating temporary directory: %w", err)
+	}
+
+	installerPath, err := r.OrbitClient.DownloadSoftwareInstaller(installer.InstallerID, tmpDir)
+	if err != nil {
+		return payload, err
+	}
+
+	// remove tmp directory and installer
+	defer func() {
+		if r.removeAllFn == nil {
+			r.removeAllFn = os.RemoveAll
+		}
+		err := r.removeAllFn(tmpDir)
+		if err != nil {
+			log.Err(err)
+		}
+	}()
+
+	installOutput, installExitCode, err := r.runInstallerScript(ctx, installer.InstallScript, installerPath, "install-script")
+	payload.InstallScriptOutput = &installOutput
+	payload.InstallScriptExitCode = &installExitCode
+	if err != nil {
+		return payload, err
+	}
+
+	if installer.PostInstallScript != "" {
+		postOutput, postExitCode, postErr := r.runInstallerScript(ctx, installer.PostInstallScript, installerPath, "post-install-script")
+		payload.PostInstallScriptOutput = &postOutput
+		payload.PostInstallScriptExitCode = &postExitCode
+
+		if postErr != nil || postExitCode != 0 {
+			log.Info().Msgf("installation of %s failed, attempting rollback. Exit code: %d, error: %s", installerPath, postExitCode, postErr)
+			uninstallScript := file.GetRemoveScript(installerPath)
+			uninstallOutput, uninstallExitCode, uninstallErr := r.runInstallerScript(ctx, uninstallScript, installerPath, "rollback-script")
+			log.Info().Msgf(
+				"rollback staus: exit code: %d, error: %s, output: %s",
+				uninstallExitCode, uninstallErr, uninstallOutput,
+			)
+
+			return payload, postErr
+		}
+	}
+
+	return payload, nil
+}
+
+func (r *Runner) runInstallerScript(ctx context.Context, scriptContents string, installerPath string, fileName string) (string, int, error) {
+	// run script in installer directory
+	installerDir := filepath.Dir(installerPath)
+	scriptPath := filepath.Join(installerDir, fileName)
+	if err := os.WriteFile(scriptPath, []byte(scriptContents), constant.DefaultFileMode); err != nil {
+		return "", -1, fmt.Errorf("writing script: %w", err)
+	}
+
+	if r.execCmdFn == nil {
+		r.execCmdFn = scripts.ExecCmd
```

Comment:
This could be a race, should do it the other way around and not alter the struct concurrently:
```
execFn := r.execCmdFn
if execFn == nil {
    execFn = scripts.ExecCmd
}
```
(and then use `execFn`)

Reply:
Done!